docs/troubleshooting/vm.md

```diff
@@ -243,3 +243,111 @@ After the pod is successfully deleted, the `Start` button becomes visible again
+cher-caddytest-9tjzj" podUID=d512bf3e-f215-4128-960a-0658f7e63c7c
+```
+
+The keywords `failed to generate spec: not a device node` are found in containerd log:
+
+```
+file path: /var/lib/rancher/rke2/agent/containerd/containerd.log
+
+time="2024-02-21T11:24:00.140298800Z" level=error msg="CreateContainer within sandbox \"850958f388e63f14a683380b3c52e57db35f21c059c0d93666f4fdaafe337e56\" for &ContainerMetadata{Name:compute,Attempt:0,} failed" error="failed to generate container \"5ddad240be2731d5ea5210565729cca20e20694e364e72ba14b58127e231bc79\" spec: failed to generate spec: not a device node"
+
+```
+
+After adding debuging information to contaierd, it shows which file is `not a device node`.
+
+```
+time="2024-02-22T15:15:08.557487376Z" level=error msg="CreateContainer within sandbox \"d23af3219cb27228623cf8168ec27e64e836ed44f2b2f9cf784f0529a7f92e1e\" for &ContainerMetadata{Name:compute,Attempt:0,} failed" error="failed to generate container \"e4ed94fb5e9145e8716bcb87aae448300799f345197d52a617918d634d9ca3e1\" spec: failed to generate spec: get device path: /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19 containerPath: /dev/disk-0 error: not a device node"
+```
+
+This is a CSI related file, but it is an empty file instead a device file, then the containerd denied the `CreateContainer` request.
+
+```
+$ ls /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/ -alth
+total 8.0K
+drwxr-x--- 2 root root 4.0K Feb 22 15:10 .
+-rw-r--r-- 1 root root    0 Feb 22 14:28 aa851da3-cee1-45be-a585-26ae766c16ca
+-rw-r--r-- 1 root root    0 Feb 22 14:07 20291c6b-62c3-4456-be8a-fbeac118ec19
+drwxr-x--- 4 root root 4.0K Feb 22 14:06 ..
+-rw-r--r-- 1 root root    0 Feb 21 15:48 4333c9fd-c2c8-4da2-9b5a-1a310f80d9fd
+-rw-r--r-- 1 root root    0 Feb 21 09:18 becc0687-b6f5-433e-bfb7-756b00deb61b
+
+$file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19
+: empty
+```
+
+By contrast, another running VM's related file is the expected device file.
+
+```
+$ ls  /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-732f8496-103b-4a08-83af-8325e1c314b7/ -alth
+total 8.0K
+drwxr-x--- 2 root root  4.0K Feb 21 10:53 .
+drwxr-x--- 4 root root  4.0K Feb 21 10:53 ..
+brw-rw---- 1 root root 8, 16 Feb 21 10:53 4883af80-c202-4529-a2c6-4e7f15fe5a9b
+```
+
+#### Root Cause
+
+The kubelet calls `NodePublishVolume` for the new pod after node/cluster rebooting without first calling `NodeStageVolume`, and the Longhorn CSI plugin bind mounts the regular file at staging target path (previously used by the deleted pod) to target path and returns success.
+
+### Workaround
+
+To address the issue, you can force delete the pod using the command `kubectl delete pod virt-launcher-*** -n namespace --force`.
```

Comment:
In my testing with a Deployment that is only allowed to run on one node, force deleting the Pod does NOT mitigate the problem. The Deployment controller reschedules the Pod to the same node and `NodePublishVolume` is again called without `NodeStageVolume`.

In addition, even though we no longer see the Pod with kubectl, kubelet can never clean up its PersistentVolume, so Longhorn can never be asked to detach it. This can lead to many additional problems down the line.

After much experimentation, see https://github.com/longhorn/longhorn/issues/8009#issuecomment-1979525625 for the workarounds that helped in the pure Longhorn (with Deployments) reproduce. I think they will be applicable in the Harvester situation as well.

Reply:
I can confirm force delete does not solve the issue. I had to restart the kubelet and then force-delete the pod to get reconcile to be retriggered.

Reply:
Sorry, I forgot to remove the workaround, now updated as `TBD`; discussed with @ejweber and are working on a solution for Harvester.

Reply:
@ejweber I copied and updated your option 2 to this PR to as a workaround for Harvester. And add you as the co-author . Thanks.